include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -536,85 +537,172 @@ struct __get_set_operation<oneapi::dpl::unseq_backend::_SymmetricDifferenceTag<s
-
-    if (__comp(__ele_val, __rng2[__merge_path_rng2]))
+    else
     {
-        // There is no chance that the balanced path differs from the merge path here, because the previous element of
-        // rng1 does not match the next element of rng2. We can just return the merge path.
-        return std::make_tuple(__merge_path_rng1, __merge_path_rng2, false);
+        return std::make_tuple(__rng1_idx, __rng2_idx);
     }
+}
 
-    // find first element of repeating sequence in the first set of the previous element
-    _Index __rng1_repeat_start = oneapi::dpl::__internal::__biased_lower_bound</*__last_bias=*/true>(
-        __rng1, _Index{0}, __merge_path_rng1, __ele_val, __comp);
-    // find first element of repeating sequence in the second set of the next element
-    _Index __rng2_repeat_start = oneapi::dpl::__internal::__biased_lower_bound</*__last_bias=*/true>(
-        __rng2, _Index{0}, __merge_path_rng2, __ele_val, __comp);
+template <typename _Rng, typename _IdxT>
+std::tuple<_IdxT, _IdxT>
+__decode_balanced_path_temp_data_no_star(const _Rng& __rng, const _IdxT __id, const std::uint16_t __diagonal_spacing)
+{
+    return __decode_balanced_path_temp_data_impl<false>(__rng, __id, __diagonal_spacing);
+}
 
-    _Index __rng1_repeats = __merge_path_rng1 - __rng1_repeat_start;
-    _Index __rng2_repeats_bck = __merge_path_rng2 - __rng2_repeat_start;
+template <typename _Rng, typename _IdxT>
+std::tuple<_IdxT, _IdxT, decltype(std::declval<_Rng>().size())>
+__decode_balanced_path_temp_data(const _Rng& __rng, const _IdxT __id, const std::uint16_t __diagonal_spacing)
+{
+    return __decode_balanced_path_temp_data_impl<true>(__rng, __id, __diagonal_spacing);
+}
 
-    if (__rng2_repeats_bck >= __rng1_repeats)
-    {
-        // If we have at least as many repeated elements in rng2, we end up back on merge path
-        return std::make_tuple(__merge_path_rng1, __merge_path_rng2, false);
-    }
+//TODO: check on types here
+template <typename _IdxT>
+auto
+__encode_balanced_path_temp_data(const _IdxT __rng1_idx, const bool __star)
+{
+    return __rng1_idx * (__star ? -1 : 1);
+}
 
-    // Calculate the number of "unmatched" repeats in the first set, add one and divide by two to round up for a
-    // possible star diagonal.
-    _Index __fwd_search_count = (__rng1_repeats - __rng2_repeats_bck + 1) / 2;
+struct __get_bounds_partitioned
+{
+    template <typename _Rng, typename _IndexT>
+    auto // Returns a tuple of the form (start1, end1, start2, end2)
+    operator()(const _Rng& __in_rng, const _IndexT __id) const
+    {
+        auto __rng_tmp_diag = std::get<2>(__in_rng.tuple()); // set a temp storage sequence
 
-    // Calculate the max location to search in the second set for future repeats, limiting to the edge of the range
-    _Index __fwd_search_bound = std::min(__merge_path_rng2 + __fwd_search_count, __rng2.size());
+        using _SizeType = decltype(std::get<0>(__in_rng.tuple()).size());
```

Comment:
No necessary action on your side here, but I see we use something along the lines of `decltype(__rng.size()))` frequently in this patch. It looks like we have a `oneapi::dpl::__internal::__range_size_t` but only for C++20 or later. Perhaps it would be worth extending this for C++17 with internal ranges.